### PR TITLE
Add support for swarm scope docker networks

### PR DESCRIPTION
**Pull request by ishantt**
This patch adds support for the swarm scope docker networks
Changes done:
*) Implement distributed lock for the shared network to prevent race
conditions
*) Fix RequestAddress to allocate gateway for the overlay networks

### CODE REVIEW

File: driver/infoblox_ipam.go [hunk 1]
```diff
@@ -56,42 +58,45 @@ func getPrefixLength(cidr string) (prefixLength string) {
...
+	var fixedAddr *ibclient.FixedAddress
 	var err error
-	if fixedAddr == nil {
+	// TODO: Do not create multiple gateways for shared networks
+
+	// Gateway request does not have a MAC address in it. If RequestAddressType is Gateway
+	// then allocate IP with MAC 00:00:00:00:00:00
+	// If RequestAddressType is not Gateway and MAC is empty then its overlay network
+	// request for the Gateway.
+	if rType, ok := r.Options["RequestAddressType"]; macAddr == "" || (ok && rType == netlabel.Gateway) {
```
Comment (yuewko) [inline driver/infoblox_ipam.go:76]: Curious...

So, there are cases where macAddr is "" even when the interface is NOT a gateway?

File: driver/lock.go [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+	logrus.Debugf("Creating lock %s\n", l.name)
+
+	eaDef := ibclient.EADefinition{Name: l.name, Type: "STRING", Flags: "C",
+		Comment:      "Docker Resource Lock",
+		DefaultValue: string(time.Now().Unix())}
+
+	// TODO: implement timeout so that lock is never held infinitely for a resource
+	// This can happen when a system which acquired lock crashes without releasing the lock
+
+	eaDefRef, err := l.objMgr.CreateEADefinition(eaDef)
```
Comment (yuewko) [inline driver/lock.go]: A couple of things:

1) Create EA Definition requires "Superuser" permission. We should try to accomplish the same using EA instead.

2) You might want to have a timestamp as part of the persisted GridLock object, so the code can recover from situation if a lock has not been released due to an exception. i.e. if the timestamp of a lock is beyond a certain age, the code can assume it's a stale lock.

File: driver/infoblox_ipam.go (hunk 1 above)
Comment (ishantt) [reply driver/infoblox_ipam.go:76]: Yes there are requests in overlay networks which does not have a MAC and are also not for the Gateway . 
https://github.com/moby/moby/issues/33692

File: driver/lock.go (hunk 2 above)
Comment (ishantt) [reply driver/lock.go]: Thanks Ko!
1) I will check on this
2) Yes that is in my todo list. I am storing  the timestamp as default value of EAdefinition https://github.com/ishantt/docker-infoblox/blob/swarm/driver/lock.go#L25

File: driver/infoblox_ipam.go (hunk 1 above)
Comment (yuewko) [reply driver/infoblox_ipam.go:76]: Thanks! Actually I remember reading about that (https://github.com/moby/moby/issues/33692). If I understand correctly, Madhu seems to be agreeing that it's a bug. So, you might need to keep track of this.

File: driver/lock.go (hunk 2 above)
Comment (yuewko) [reply driver/lock.go]: Perfect!!

File: driver/infoblox_ipam.go [hunk 3]
```diff
@@ -200,50 +187,142 @@ func (ibDrv *InfobloxDriver) allocateNetwork(netview string, prefixLen uint, net
...
+		// Network exists and is same as requested
+		logrus.Debugf("Shared network (%s) found with CIDR (%s)\n", networkName, networkByName.Cidr)
+		return networkByName, nil
+	}
+	return networkByName, nil
+}
+
+func (ibDrv *InfobloxDriver) createSharedNetwork(netViewName string, pool string, networkName string, prefixLen uint) (*ibclient.Network, error) {
+
+	l := NVLocker{name: netViewName, objMgr: ibDrv.objMgr}
```
Comment (yuewko) [inline driver/infoblox_ipam.go]: Recommend implementing a factory method for NVLocker.

File: driver/ipam_driver.go [hunk 4]
```diff
@@ -106,4 +106,5 @@ func main() {
 
 	h := ipamPluginSdk.NewHandler(ipamDrv)
 	h.ServeUnix(socketAddress, 0)
+
```
Comment (yuewko) [inline driver/ipam_driver.go]: Newline added by accident?

File: driver/lock.go [hunk 5]
```diff
@@ -0,0 +1,194 @@
+package main
+
```
Comment (yuewko) [inline driver/lock.go]: Overall comment...

Have you considered making this part of Infoblox-go-client? IOW, do you think the NVLocker can be of more general use? For example, would this be useful for the CNI plugin too?

File: driver/infoblox_ipam.go [hunk 6]
```diff
@@ -200,50 +187,142 @@ func (ibDrv *InfobloxDriver) allocateNetwork(netview string, prefixLen uint, net
...
+			// sleep for random time (between 1 - 5 seconds) to reduce collisions
+			time.Sleep(time.Duration(rand.Intn(4)+1) * time.Second)
+			continue
+		} else {
+			// Got the lock. Create the network.
+			logrus.Debugf("Got the lock. Creating the network %s\n", networkName)
+			defer l.UnLock(false)
+			break
+		}
+	}
```
Comment (yuewko) [inline driver/infoblox_ipam.go]: Have you considered making the above a generic function of the NVLocker class?

**Review by yuewko — COMMENTED** [review]
Looking good!

A few comments inline.

Thanks!

File: driver/infoblox_ipam.go (hunk 6 above)
Comment (yuewko) [reply driver/infoblox_ipam.go]: To clarify, the above comments refers to L222-L243

File: driver/infoblox_ipam.go (hunk 3 above)
Comment (ishantt) [reply driver/infoblox_ipam.go]: Ok Thanks!

File: driver/infoblox_ipam.go (hunk 6 above)
Comment (ishantt) [reply driver/infoblox_ipam.go]: Yeah Nice Idea. Will make this function clean. Thanks!

File: driver/ipam_driver.go (hunk 4 above)
Comment (ishantt) [reply driver/ipam_driver.go]: oops..I will remove it.

File: driver/lock.go (hunk 5 above)
Comment (ishantt) [reply driver/lock.go]: Yes makes sense. For CNI in kubernetes etcd can also be used for locking.

**Review by yuewko — APPROVED** [review]
LGTM

**Pull request merged**